ibis/expr/types/relations.py

```diff
@@ -2911,29 +2935,20 @@ def join(
         │  106782 │ Leonardo DiCaprio │          5989 │ Leonardo DiCaprio │
         └─────────┴───────────────────┴───────────────┴───────────────────┘
         """
+        from ibis.expr.types.joins import JoinExpr
 
-        _join_classes = {
-            "inner": ops.InnerJoin,
-            "left": ops.LeftJoin,
-            "any_inner": ops.AnyInnerJoin,
-            "any_left": ops.AnyLeftJoin,
-            "outer": ops.OuterJoin,
-            "right": ops.RightJoin,
-            "left_semi": ops.LeftSemiJoin,
-            "semi": ops.LeftSemiJoin,
-            "anti": ops.LeftAntiJoin,
-            "cross": ops.CrossJoin,
-        }
-
-        klass = _join_classes[how.lower()]
-        expr = klass(left, right, predicates).to_expr()
+        # the first participant of the join shouldn't be a self reference
+        left = left.op()
+        if isinstance(left, ops.SelfReference):
+            left = left.parent
 
-        # semi/anti join only give access to the left table's fields, so
-        # there's never overlap
-        if how in ("left_semi", "semi", "anti"):
-            return expr
+        # construct an empty join chain and wrap it with a JoinExpr
```

Comment:
This comment doesn't seem that helpful, it's describing the code below which isn't obfuscated in any way by indirection.